### PR TITLE
C++ & JS via JSON

### PR DESCRIPTION
**Pull request by tdresser**

### CODE REVIEW

**Review by ojanvafai — APPROVED** [review]
Why switch the build file to js?

File: cpp/main.cc [hunk 1]
```diff
@@ -14,31 +15,98 @@ string getStringCC() {
...
+  p.to_json(j);
+}
+
+void from_json(const json& j, Point& p) {
+  p.from_json(j);
+}
+
+enum class ActionType { Unknown, TouchStart, TouchMove };
+
+NLOHMANN_JSON_SERIALIZE_ENUM(ActionType,
```
Comment (ojanvafai) [inline cpp/main.cc:48]: Lol, we should probably typedef NLOHMANN_JSON_SERIALIZE_ENUM at some point.

File: cpp/main.cc [hunk 2]
```diff
@@ -14,31 +15,98 @@ string getStringCC() {
...
  public:
-  Point(float x, float y) {
-    PointS::x = x;
-    PointS::y = y;
-  };
+  Action(ActionType type, Point point) : _type(type), _point(point) {}
+  Action(){};
+  ActionType type() const { return _type; };
+  const Point& point() const { return _point; }
+  void to_json(json& j) const { j = json{{"type", _type}, {"point", _point}}; }
```
Comment (ojanvafai) [inline cpp/main.cc:61]: Can we use camelCase? I know google style is snake_case, but I can't really stand it.

When you break it up into files, can we use FullCaps.h case?

I like that searching over the codebase naturally matches both the filename and the exported class name and that we don't need to search with different casing for the JS and C++ sides of the codebase.

File: cpp/main.cc [hunk 3]
```diff
@@ -14,31 +15,98 @@ string getStringCC() {
...
   return 7;
 }
 
 EMSCRIPTEN_BINDINGS(geometry) {
-  value_object<PointS>("Point").field("x", &PointS::x).field("y", &PointS::y);
-
-  function<PointS>("getCardOffset", &getCardOffset);
+  function<string>("getCardOffset", &getCardOffset);
   function<int>("testGetString", &testGetString);
 };
```
Comment (ojanvafai) [inline cpp/main.cc:112]: WDYT of adding `"files.insertFinalNewline": true` to our workspace settings.json so we don't have to remember to put trailing new lines?

**Comment by tdresser** [comment]
Doing these in the next patch.

**Pull request merged**